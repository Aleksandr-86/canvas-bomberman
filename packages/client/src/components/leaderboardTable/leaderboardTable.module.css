.table {
  border-collapse: collapse;
  font-size: 40px;
  user-select: none;
}

.caption {
  margin-bottom: 30px;
  font-size: 40px;
  text-transform: uppercase;
}

.head {
  display: flex;
  padding-bottom: 10px;
  cursor: pointer;
}

.title {
  display: flex;
  justify-content: center;
}

.ascending::after {
  content: '▲';
  display: inline-block;
  margin-left: 10px;
}

.descending::after {
  content: '▼';
  display: inline-block;
  margin-left: 10px;
}

.paddingBorder {
  padding: 10px 0;
  border-bottom: 2px solid #ccc;
}

.columnPlace {
  width: 220px;
  text-align: center;
}

.columnName {
  width: 420px;
  text-align: left;
}

.columnScore {
  width: 200px;
  text-align: center;
}

.columnGames {
  width: 330px;
  text-align: center;
}

.body {
  height: 345px;
  border-top: 5px solid black;
}

.row {
  display: flex;
  flex-direction: row;
}

.row:hover {
  background: hsl(29deg 6% 74%);
  transition: all 0.2s linear;
}
