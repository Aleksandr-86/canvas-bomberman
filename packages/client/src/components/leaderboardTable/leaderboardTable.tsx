import React from 'react'
import { useSortableData } from '../../hooks/userSortableData'
import { PlayerStats } from '../../store/leadeboard/leaderboardSlice'
import styles from './leaderboardTable.module.css'

interface Props {
  state: PlayerStats[]
}

export const LeaderboardTable: React.FC<Props> = props => {
  const { items, requestSort, sortConfig } = useSortableData(props.state, {
    key: 'place',
    direction: 'ascending',
  })

  const getClassNamesFor = (name: string) => {
    if (!sortConfig) {
      return
    }

    const { key, direction } = sortConfig

    if (key === name) {
      return `${styles.title} ${styles[direction]}`
    } else {
      return styles.title
    }
  }

  return (
    <div className={styles.table}>
      <div className={styles.head}>
        <div
          onClick={() => requestSort('place')}
          className={styles.columnPlace + ' ' + getClassNamesFor('place')}>
          Место
        </div>
        <div
          onClick={() => requestSort('name')}
          className={styles.columnName + ' ' + getClassNamesFor('name')}>
          Имя игрока
        </div>
        <div
          onClick={() => requestSort('score')}
          className={styles.columnScore + ' ' + getClassNamesFor('score')}>
          Очки
        </div>
        <div
          onClick={() => requestSort('games')}
          className={styles.columnGames + ' ' + getClassNamesFor('games')}>
          Игр сыграно
        </div>
      </div>

      <div className={styles.body}>
        {items.map(item => (
          <div className={styles.row} key={item.id}>
            <div className={styles.columnPlace + ' ' + styles.paddingBorder}>
              {item.place}
            </div>
            <div className={styles.columnName + ' ' + styles.paddingBorder}>
              {item.name}
            </div>
            <div className={styles.columnScore + ' ' + styles.paddingBorder}>
              {item.score}
            </div>
            <div className={styles.columnGames + ' ' + styles.paddingBorder}>
              {item.games}
            </div>
          </div>
        ))}
      </div>
    </div>
  )
}
