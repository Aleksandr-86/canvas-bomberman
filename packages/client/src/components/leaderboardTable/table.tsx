import { useSortableData } from '../../hooks/userSortableData'

import styles from './table.module.css'

interface Props {
  playersStats: {
    id: number
    place: number
    name: string
    score: number
    games: number
    totalTime: number
  }[]
}

const headers = [
  ['place', 'Место'],
  ['name', 'Имя игрока'],
  ['score', 'Очки'],
  ['games', 'Игр'],
  ['totalTime', 'Время (мин)'],
]

export const Table = (props: Props) => {
  const { items, requestSort, sortConfig } = useSortableData(
    props.playersStats,
    {
      key: 'place',
      direction: 'ascending',
    }
  )

  const getClassNamesFor = (name: string) => {
    if (!sortConfig) {
      return
    }

    const { key, direction } = sortConfig

    if (key === name) {
      return `${styles.head} ${styles[direction]}`
    } else {
      return styles.head
    }
  }

  return (
    <table className={styles.table}>
      <thead>
        <tr>
          {headers.map(([headerName, headerTitle], index) => (
            <th
              key={index}
              onClick={() => requestSort(headerName)}
              className={getClassNamesFor(headerName)}>
              {headerTitle}
            </th>
          ))}
        </tr>
      </thead>

      <tbody>
        {items.map(item => (
          <tr className={styles.row} key={item.id}>
            <td className={styles.columnPlace}>{item.place}</td>
            <td className={styles.columnName}>{item.name}</td>
            <td className={styles.columnScore}>{item.score}</td>
            <td className={styles.columnGames}>{item.games}</td>
            <td className={styles.columnTotalTime}>{item.totalTime}</td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
