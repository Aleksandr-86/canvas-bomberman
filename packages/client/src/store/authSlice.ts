import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { useAuth } from '../hooks/useAuth'
import { IFormData } from '../hooks/useAuth'
import { UserState } from './userSlice'

export const getAuth = createAsyncThunk(
  'user/getAuth',
  async (FormData: IFormData) => useAuth(FormData)
)

const state: UserState = {
  isLoading: false,
  isAuth: false,
  id: false,
  error: false,
  firstName: '',
  secondName: '',
  displayName: 'Алекс',
  login: '',
  email: '',
  phone: '',
}

const authSlice = createSlice({
  name: 'user',
  initialState: state,
  reducers: {},
  extraReducers: builder => {
    // Add reducers for additional action types here, and handle loading state as needed
    builder
      .addCase(getAuth.pending, state => {
        state.error = false
        state.isAuth = false
        state.isLoading = true
      })
      .addCase(getAuth.fulfilled, (state, action) => {
        const { user } = action.payload
        state.isLoading = false
        state.isAuth = user.isAuth
        if (user.isAuth) {
          state.id = user.id
          state.firstName = user.firstName
          state.secondName = user.secondName
          state.displayName = user.displayName
          state.login = user.login
          state.email = user.email
          state.phone = user.phone
        }
      })
      .addCase(getAuth.rejected, state => {
        state.isLoading = false
        state.isAuth = false
        state.error = true
      })
  },
})

const { reducer } = authSlice
export const authReducer = reducer
