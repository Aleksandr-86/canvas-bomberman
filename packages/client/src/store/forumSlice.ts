import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import axios, { AxiosError } from 'axios'

export const getPosts = createAsyncThunk('forum/getPosts', async () => {
  try {
    const response = await axios.get(
      'https://alt-f4-bomberman-21.ya-praktikum.tech/posts'
    )
    return response.data
  } catch (e) {
    return e as AxiosError
  }
})

export const sendPost = createAsyncThunk(
  'users/sendPost',
  async (formData: FormData) => {
    try {
      const response = await axios.post(
        'https://alt-f4-bomberman-21.ya-praktikum.tech/posts',
        {
          data: formData,
        }
      )
      return response.data
    } catch (e) {
      return e as AxiosError
    }
  }
)

export const sendLike = createAsyncThunk(
  'users/sendPost',
  async (id: number) => {
    try {
      await axios.post(
        'https://alt-f4-bomberman-21.ya-praktikum.tech/likes',
        {
          data: id,
        }
      )
      return id
    } catch (e) {
      return e as AxiosError
    }
  }
)

export const sendDislike = createAsyncThunk(
  'users/sendPost',
  async (id: number) => {
    try {
      await axios.post(
        'https://alt-f4-bomberman-21.ya-praktikum.tech/dislikes',
        {
          data: id,
        }
      )
      return id
    } catch (e) {
      return e as AxiosError
    }
  }
)

export const getMessages = createAsyncThunk('forum/getMessages', async id => {
  try {
    const response = await axios.get(
      `https://alt-f4-bomberman-21.ya-praktikum.tech/posts/${id}`
    )
    return response.data
  } catch (e) {
    return e as AxiosError
  }
})

export const sendMessage = createAsyncThunk(
  'users/sendMessage',
  async (formData: FormData) => {
    try {
      const response = await axios.post(
        `https://alt-f4-bomberman-21.ya-praktikum.tech/posts/`,
        {
          data: formData,
        }
      )
      return response.data
    } catch (e) {
      return e as AxiosError
    }
  }
)

const FORUM_DATA = {
  headers: ['Автор', 'Тема', 'Дата'],
  rows: [
    {
      id: 1,
      author: 'Kan',
      body: 'Как пройти первый уровень',
      date: '8.01.22',
    },
    {
      id: 2,
      author: 'sasha',
      body: 'Как создать новую тему',
      date: '8.01.22',
    },
  ],
  redirect: true
}
const FORUM_MESSAGES = {
  headers: ['Автор', 'Сообщения', 'Дата', 'Лайки', 'Дизлайки'],
  rows: [
    {
      id: 1,
      postid: 1,
      author: 'Kan88',
      body: 'Как пройти первый уровень',
      date: '8.01.22',
      likes: 150,
      dislikes: 12
    },
    {
      id: 2,
      postid: 1,
      author: 'sasha85',
      body: 'Я прошел на изи, надо лишь быстро двигать пальцами',
      date: '8.01.22',
      likes: 2,
      dislikes: 12
    }
  ],
}

const initialState = {
  posts: FORUM_DATA,
  messages: FORUM_MESSAGES,
  error: null as unknown | null | string,
}

const forumSlice = createSlice({
  name: 'users',
  initialState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(getPosts.pending, state => {
      state.error = null
    })
    builder.addCase(sendPost.pending, state => {
      state.error = null
    })
    builder.addCase(sendLike.pending, state => {
      state.error = null
    })
    builder.addCase(sendDislike.pending, state => {
      state.error = null
    })
    builder.addCase(getMessages.pending, state => {
      state.error = null
    })
    builder.addCase(sendMessage.pending, state => {
      state.error = null
    })
    builder.addCase(getPosts.fulfilled, (state, action) => {
      state.posts = { ...action.payload }
    })
    builder.addCase(sendPost.fulfilled, (state, action) => {
      state.posts.rows.push(action.payload)
    })
    builder.addCase(sendLike.fulfilled, (state, action) => {
      state.messages.rows.map((item) => {
        if (item.id === action.payload) {
          item.likes++
        }
      })
    })
    builder.addCase(sendDislike.fulfilled, (state, action) => {
      state.messages.rows.map((item) => {
        if (item.id === action.payload) {
          item.dislikes++
        }
      })
    })
    builder.addCase(getMessages.fulfilled, (state, action) => {
      state.messages = { ...action.payload }
    })
    builder.addCase(sendMessage.fulfilled, (state, action) => {
      state.messages.rows.push(action.payload)
    })
    builder.addCase(getPosts.rejected, (state, action) => {
      state.error = action.payload
    })
    builder.addCase(sendPost.rejected, (state, action) => {
      state.error = action.payload
    })
    builder.addCase(sendLike.rejected, (state, action) => {
      state.error = action.payload
    })
    builder.addCase(sendDislike.rejected, (state, action) => {
      state.error = action.payload
    })
    builder.addCase(getMessages.rejected, (state, action) => {
      state.error = action.payload
    })
    builder.addCase(sendMessage.rejected, (state, action) => {
      state.error = action.payload
    })
  },
})

export const forumReducer = forumSlice.reducer
