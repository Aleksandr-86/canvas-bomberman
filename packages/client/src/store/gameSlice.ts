import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import {
  PLAYER_STARTING_POSITION,
  EMPTY_FIELD,
  GRID_HEIGHT,
  GRID_WIDTH,
} from '../features/game/const'
import { Kind, type FieldState } from '../features/game/types'
import { delay, Vec2 } from '../features/game/utils'
import { fillSoftWalls } from '../features/game/fillSoftWalls'
import { explodeAdjacentSoftWalls } from '../features/game/explodeAdjacentSoftWalls'
import { AppDispatch } from './index'

export enum GameStatus {
  START = 'START',
  IN_PROGRESS = 'IN_PROGRESS',
  END = 'END',
}

type GameState = {
  field: FieldState
  status: GameStatus
  currentScore: number
  playerPosition: Vec2
  doorPosition: Vec2
  bombs: Vec2[]
  time: number
}

const INITIAL_STATE: GameState = {
  status: GameStatus.START,
  currentScore: 0,
  doorPosition: new Vec2(GRID_WIDTH - 2, GRID_HEIGHT - 2),
  playerPosition: PLAYER_STARTING_POSITION,
  bombs: [],
  field: EMPTY_FIELD,
  time: 300,
}

const gameSlice = createSlice({
  name: 'game',
  initialState: () => {
    return {
      ...INITIAL_STATE,
      field: fillSoftWalls(EMPTY_FIELD),
    }
  },
  reducers: {
    setStatus: (state, action: PayloadAction<GameStatus>) => {
      state.status = action.payload
    },
    setCurrentScore: (state, action: PayloadAction<number>) => {
      state.currentScore = action.payload
    },
    playerMoved: (state, { payload }: PayloadAction<Vec2>) => {
      state.playerPosition = payload
    },
    playerReset: state => {
      state.playerPosition = PLAYER_STARTING_POSITION
    },
    bombPlaced: (state, { payload: { x, y } }: PayloadAction<Vec2>) => {
      state.field[y][x] = Kind.Bomb
    },
    scoreIncreased: (state, { payload }: PayloadAction<number>) => {
      state.currentScore += payload
    },
    bombExploded: ({ field }, { payload }: PayloadAction<Vec2>) => {
      explodeAdjacentSoftWalls(field, payload)
    },
    timerUpdated: (state, { payload }: PayloadAction<number>) => {
      state.time -= payload * 0.001 /* payload is in ms */
    },
  },
})

export const bombSet = (pos: Vec2) => async (dispatch: AppDispatch) => {
  dispatch(bombPlaced(pos))
  await delay(800)
  dispatch(bombExploded(pos))
}

let previousTime = performance.now()
let gameIntervalId = 0

export const gameStarted = () => (dispatch: AppDispatch) => {
  gameIntervalId = window.setInterval(() => {
    const now = performance.now()
    const dt = now - previousTime
    previousTime = now
    dispatch(timerUpdated(dt))
  }, 1000)
  dispatch(setStatus(GameStatus.IN_PROGRESS))
}

export const gameEnded = () => (dispatch: AppDispatch) => {
  clearInterval(gameIntervalId)
  dispatch(setStatus(GameStatus.END))
}

export const gameReducer = gameSlice.reducer
export const {
  setStatus,
  setCurrentScore,
  playerMoved,
  playerReset,
  bombPlaced,
  bombExploded,
  scoreIncreased,
  timerUpdated,
} = gameSlice.actions
