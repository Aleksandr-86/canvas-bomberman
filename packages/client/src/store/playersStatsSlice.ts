import { createSlice } from '@reduxjs/toolkit'

export type PlayerStats = {
  id: number
  place: number
  name: string
  score: number
  games: number
  totalTime: number
}

type PlayersStatsState = PlayerStats[]

const initialState: PlayersStatsState = [
  {
    id: 1,
    place: 1,
    name: 'Bionic',
    score: 27500,
    games: 10,
    totalTime: 61,
  },
  {
    id: 2,
    place: 2,
    name: 'Chrome Road',
    score: 15150,
    games: 5,
    totalTime: 36,
  },
  {
    id: 3,
    place: 3,
    name: 'Bladewatch',
    score: 13500,
    games: 4,
    totalTime: 30,
  },
  {
    id: 4,
    place: 4,
    name: 'Marshmallow',
    score: 12100,
    games: 3,
    totalTime: 23,
  },
  {
    id: 5,
    place: 5,
    name: 'Arctic Realm',
    score: 9800,
    games: 3,
    totalTime: 19,
  },
]

const playersStatsSlice = createSlice({
  name: 'playersStats',
  initialState,
  reducers: {},
})

export const playersStatsReducer = playersStatsSlice.reducer
