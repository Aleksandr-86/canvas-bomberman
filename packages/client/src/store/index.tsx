import { configureStore } from '@reduxjs/toolkit'
import { userReducer } from './reducers/userSlice'
import { leaderboardReducer } from './reducers/leaderboardSlice'
import { gameReducer } from './reducers/gameSlice'

export const store = configureStore({
  reducer: {
    user: userReducer,
    leaderboard: leaderboardReducer,
    game: gameReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
