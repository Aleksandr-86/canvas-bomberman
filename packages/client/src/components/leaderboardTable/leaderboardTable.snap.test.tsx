import * as renderer from 'react-test-renderer'
import { LeaderboardTable } from './leaderboardTable'

import { type PlayerStats } from '../../store/playersStatsSlice'

const mockProps: PlayerStats[] = [
  {
    id: 1,
    place: 1,
    name: `Alex`,
    score: 811,
    games: 1,
  },
]

describe(`<LeaderboardTable/> snapshot test cases`, () => {
  test(`<LeaderboardTable/> should render <table className=table> parent with child elements (with mockProps)`, () => {
    const tree = renderer
      .create(<LeaderboardTable state={mockProps} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
