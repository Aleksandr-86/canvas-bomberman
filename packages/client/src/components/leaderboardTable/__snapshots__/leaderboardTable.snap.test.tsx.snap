// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LeaderboardTable/> snapshot test cases <LeaderboardTable/> should render <table className=table> parent with child elements (with mockProps) 1`] = `
<div
  className="table"
>
  <div
    className="head"
  >
    <div
      className="columnPlace title ascending"
      onClick={[Function]}
    >
      Место
    </div>
    <div
      className="columnName title"
      onClick={[Function]}
    >
      Имя игрока
    </div>
    <div
      className="columnScore title"
      onClick={[Function]}
    >
      Очки
    </div>
    <div
      className="columnGames title"
      onClick={[Function]}
    >
      Игр сыграно
    </div>
  </div>
  <div
    className="body"
  >
    <div
      className="row"
    >
      <div
        className="columnPlace paddingBorder"
      >
        1
      </div>
      <div
        className="columnName paddingBorder"
      >
        Alex
      </div>
      <div
        className="columnScore paddingBorder"
      >
        811
      </div>
      <div
        className="columnGames paddingBorder"
      >
        1
      </div>
    </div>
  </div>
</div>
`;
